Bring calculations into better alignment with RFC.

beevik opened the pull request:
RTT calculation can no longer yield a result less than the precision value.
Account for minimum dispersion in the calculation of the root distance.
Updated the comments for all calculated fields to better match the RFC.
Removed casuality violation until I have a better understanding of how it's useful.

beevik commented:
@darkk, could you please review this submission?  Among other changes, I've removed the causality violation calculations, because I don't really understand how one is expected to use it or why it reveals a useful result.

darkk commented on ntp.go:
```diff
@@ -407,78 +409,57 @@ func parseTime(m *msg, recvTime ntpTime) *Response {
+		rtt = toInterval(prec)
+	}
+	return rtt
 }
```
IMHO, it's incorrect to truncate RTT to precision. Local clock may be precise enough and precision indicates some sort of remote server clock precision.

beevik replied:
This is what the reference implementation in the RFC does.

beevik replied:
I checked the reference implementation again and realized it is clamping RTT to the local system's clock precision, not the remote server's clock precision.  Since I don't know the local system's clock precision, I decided to clamp the value to zero instead.

darkk commented on ntp.go:
```diff
@@ -407,78 +409,57 @@ func parseTime(m *msg, recvTime ntpTime) *Response {
 
 	// Calculate values depending on other calculated values
 	r.RootDistance = rootDistance(r.RTT, r.RootDelay, r.RootDispersion)
-	r.CausalityViolation = causalityViolation(r.RTT, r.ClockOffset)
```
I'm sad to see you dropping `CausalityViolation` when it's used in [prometheus/node_exporter](https://github.com/prometheus/node_exporter/blob/master/collector/ntp.go#L155) :-/

beevik replied:
I'm happy to add it back in.  I'd just like to have a better understanding of what it's trying to do.  How does one interpret the resulting value?  When is a value considered "bad"?

darkk replied:
I'll try do draw some diagram explaining this value.

darkk replied:
```
S: --[t1]-[t2]---------->
       \   \__ 
        \     \__ 
         \       \   
C: ------[t0]----[t3]--->
```
That's causality violation case. If clocks are in-sync then T1 is always greater then T2.
There is Offset that is estimate of clock error that assumes that network path is symmetric in terms of latency.
And CausalityViolation is lower bound of clock error, if T0 is greater than T1 then clocks are **at least** wrong by the difference and that's 100% true :)

```
S: ------[t1]-[t2]--->
         _/    |
       _/      | 
      /       /    
C: --[t0]----[t3]---->
```
That's another example. If we assume that server clock is "true" clock then packet arrives "earlier" than departs, so we have another esimate for minimum error value.

These error values are useful to check if local kernel wall clock is synced to local NTP daemon clock in case of `chrony` daemon or external NTP server providing "reference" clock. Offset is always non-zero due to network fluctuations and CausalityViolation may give lower-bound estimate in case of unsync clock.

I hope it helps.

beevik replied:
Ok, that adds clarity.

So what you're saying is that if `CV > 0`, then the client and server clocks are not just out of sync (i.e., `ClockOffset != 0`); they are also returning values that really should be impossible.  Is that correct?

darkk replied:
> are not just out of sync (i.e., `ClockOffset != 0`);

Estimated offset (ClockOffset) is always non-zero, but 5ms offset estimated using nearby server (1ms away) will show some CV, while remote server (RTT~100ms) will show CV=0 as there is some uncertainty due to processing time.

> I don't really understand how the "causality violation" happens in the first place

Server and Client clocks have some offset between them :) Let me show an example
- client sends {0, x, x} at C:0, S:5 (true offset is 5ms)
- server receives {0, 6, x} at C:1, S:6 (server is 1ms away)
- server sends {0, 6, 6} at C:1, S:6 (no processing time)
- client receives {0, 6, 6} at C:3, S:8 (reverse path is a bit longer)
So RTT = 3, ClockOffset = +/-4.5, CV = 3. Estimated clock offset is 4.5ms, but it's >= 3ms for sure.

Does it help? I'm unsure if I understand your question correctly :-)

beevik replied:
Thanks for putting up with all my questions.

After your most recent example, I understand better the meaning of "causality violation" and how it can be greater than zero in realistic scenarios.  Maybe the name "causality violation" is what is throwing me off, because it implies something that should be impossible.  There are no physical phenomena that violate the principle of causality. 

What this really is is a lower bound on the clock offset.  Once you know this value, how is it useful?  What can you do with it?

darkk commented on ntp.go:
```diff
@@ -407,78 +409,57 @@ func parseTime(m *msg, recvTime ntpTime) *Response {
 	}
-	return time.Duration(0)
+	return totalDelay/2 + rootDisp
 }
```
It does not handle negative RTT that may theoretically happen if clock frequency is really mad.

darkk commented:
I've also commented some individual commits that looked strange to me, I'm listing them here to keep track of them :-)
- a9265d33efe202c5fa874d24e2d8f2083a291ab0
- 4d0b133f775a2eb62587496741599e6156cee400
- 341fc19de1ed3a1200e0e7968fcf4d5c6a059345
- 8c5337d9c2bb6d7c34f7774fb1ec516b6c96de8d
- fa8ab3f24bce6cb14434023670ac15100bc399cc
- ee70b4de6106653c9604cb336d676dfe00db3e6d
- 4e2be93d4ea7f6bef2a1ad7d0c94790dbb8789a5

darkk commented on ntp.go:
```diff
@@ -20,17 +20,12 @@ import (
+	MaxStratum = 16
+
+	// MaxRootDistance is the largest possible root distance (in seconds).
+	MaxRootDistance = 2.96608
```
Is it intentional to have unused constant like that?

beevik replied:
no, this should be removed.  I was originally going to add it to Validate, but decided against it.

beevik commented:
Thanks for all your help, @darkk. I think your feedback has helped me to produce a better package.

Please feel free to comment on the change I made to `CausalityViolation`. (I renamed it to `MinError` and changed the way it's computed.  It should still result in the same value as before.)

darkk commented:
> changed the way it's computed

Yep, that's wise. Old way to compute it with Offset and RTT is kinda creepy IF the decision is made to store the value as a member of struct instead of computable value :-)
